dvc/remote/s3.py

```diff
@@ -302,3 +304,57 @@ def walk_files(self, path_info, max_items=None):
                 continue
 
             yield path_info.replace(path=fname)
+
+    def _append_aws_grants_to_extra_args(self, config):
+        # Keys for extra_args can be one of the following list:
+        # https://boto3.amazonaws.com/v1/documentation/api/latest/reference/customizations/s3.html#boto3.s3.transfer.S3Transfer.ALLOWED_UPLOAD_ARGS
+        """
+          ALLOWED_UPLOAD_ARGS = [
+            'ACL', 'CacheControl', 'ContentDisposition', 'ContentEncoding',
+            'ContentLanguage', 'ContentType', 'Expires', 'GrantFullControl',
+            'GrantRead', 'GrantReadACP', 'GrantWriteACP', 'Metadata',
+            'RequestPayer', 'ServerSideEncryption', 'StorageClass',
+            'SSECustomerAlgorithm', 'SSECustomerKey', 'SSECustomerKeyMD5',
+            'SSEKMSKeyId', 'WebsiteRedirectLocation'
+          ]
+        """
+
+        grants = {
+            Config.SECTION_AWS_GRANT_FULL_CONTROL: {
+                "extra_args_key": "GrantFullControl",
+                "self_attr": "grant_full_control",
+            },
+            Config.SECTION_AWS_GRANT_READ: {
+                "extra_args_key": "GrantRead",
+                "self_attr": "grant_read",
+            },
+            Config.SECTION_AWS_GRANT_READ_ACP: {
+                "extra_args_key": "GrantReadACP",
+                "self_attr": "grant_read_acp",
+            },
+            Config.SECTION_AWS_GRANT_WRITE_ACP: {
+                "extra_args_key": "GrantWriteACP",
+                "self_attr": "grant_write_acp",
+            },
+        }
+        overwriting_acl_grantees = False
+        for grant_option, grant_keys in grants.items():
+            grant_value = config.get(grant_option, "")
+            if grant_value:
+                overwriting_acl_grantees = True
+                self.extra_args[grant_keys["extra_args_key"]] = grant_value
+                setattr(self, grant_keys["self_attr"], grant_value)
+
+        if overwriting_acl_grantees:
+            warning_message = (
+                "Using one of the [{0}] will overwrite default ACL "
+                "Grantees list. Default ACL Grantees list include Owner "
+                "with FULL_CONTROL permissions.".format(", ".join(grants))
+            )
+
+            logger.warning(warning_message)
+
+            if self.acl:
+                logger.error(
+                    "extra_args' 'acl' and grants are mutually exclusive"
+                )
```

Comment:
Maybe we should raise ConfigError here instead?

Reply:
I agree, just didn't know about `ConfigError`.

However I run out of ideas how to fix codeclimate issue 🙁

Reply:
@rbvesb Codeclimate and DeepSource are purely advisory, you don't have to make it all pass. Just be sure to look at what it is complaining about to at least acknowledge it. 🙂

Reply:
I'll try again tomorrow - now DeepSource started to fail with something unchanged :(

Reply:
@rbvesb Checked both. Both are fine, you don't need to change anything. 🙂